blog: Running DDEV in CI with WarpBuild, add mermaid diagram support (#515)
ddev/ddev.com: src/content/blog/ddev-ci-warpbuild.md

---
title: "Using WarpBuild to speed up DDEV in CI"
pubDate: 2026-01-01
modifiedDate: 2026-01-03
modifiedComment: ""
summary: Use WarpBuild to speed up DDEV in CI.
author: Andrew Berry
featureImage:
  src: /img/blog/2026/01/warp-speed-ci.jpg
  srcDark: /img/blog/2026/01/warp-speed-ci.jpg
  alt: Warp Speed CI
  caption:
  credit:
categories:
  - DevOps
  - Performance
---

> For most developers, DDEV solves a common challenge: making sure that each developer as a consistent, stable local environment for building their web application. At Lullabot, as we had more and more success with DDEV, another related issue kept coming up: how do we grow and develop our use of continuous integration and automated testing while avoiding similar challenges?

Reviewer: ```suggestion
For most developers, DDEV solves a common challenge: making sure that each developer has a consistent, stable local environment for building their web application. At Lullabot, as we had more and more success with DDEV, but another related issue kept coming up: how do we grow and develop our use of continuous integration and automated testing while avoiding similar challenges?
```

"how do we grow and develop our use of continuous integration and automated testing while avoiding similar challenges" seems pretty vague, can you make that stronger?

> A typical CI/CD pipeline is implemented using the tools and systems provided by the CI service itself. For example, at a basic level you can place shell commands inside configuration files to run tests and tools. Running those commands locally in DDEV is possible, but it's a painful copypaste process. If you're a back-end or DevOps engineer, odds are high you've wasted hours trying to figure out why a test you wrote locally isn't passing in CI – or vice versa!

Reviewer: ```suggestion
A typical CI/CD pipeline is implemented using the tools and systems provided by the CI service itself. For example, at a basic level you can place shell commands inside configuration files to run tests and tools. Running those commands locally in DDEV is possible, but it's a painful copy/paste process. If you're a back-end or DevOps engineer, odds are high you've wasted hours trying to figure out why a test you wrote locally isn't passing in CI – or vice versa!
```

The runs-on statement:

1. [Skips snapshots via commit messages](https://github.com/WarpBuilds/snapshot-save#conditional-snapshot-usage).
2. Uses a "16x" sized runner so we can run tests in parallel.
3. Creates a snapshot key with the project name, the ddev version, a manual version number, and the short hash of the files from above.

We also switch to the WarpBuild cache (so it's local to the runner) and check out the project. Update the cache paths as appropriate for your project.

```yaml
jobs:
  # other jobs...
  build-and-test:
    needs: [determine-snapshot]
    runs-on:
      "${{ contains(github.event.head_commit.message, '[warp-no-snapshot]') &&
      'warp-ubuntu-2404-x64-16x' ||
      format('warp-ubuntu-2404-x64-16x;snapshot.key=my-project-ddev-1.25.0-v1-{0}', needs.determine-snapshot.outputs.snapshot) }}"

    steps:
      - uses: WarpBuilds/cache@v1
        with:
          path: |
            ${{ github.workspace }}/.ddev/.drainpipe-composer-cache
            ${{ github.workspace }}/vendor
            ${{ github.workspace }}/web/core

>             ${{ github.workspace }}/web/modules/contrib

Author: We typically run `npm ci` to ensure consistency which means the first thing it does is remove the node_modules folder.

But, we could probably cache the downloads. In practice this section ends up not mattering much because they get saved into the snapshot, so it's only for fresh builds that it helps.

      - name: Clean up for snapshot
        if: ${{ steps.find-ddev.outputs.ddev-path != '/usr/bin/ddev' }}
        run: |
          # Stop ddev to ensure clean state
          ddev poweroff
          # Remove any cached credentials or tokens
          rm -f ~/.terminus/cache/session
          # Clean git state and temporary files
          git clean -ffdx
```

Now we can actually save the snapshot. We skip this if we can since it takes a bit of time to save and upload. There's no point in rewriting our snapshot if it hasn't changed! The `wait-timeout-minutes` is set very high, but in practice this step only takes a minute or two. We just don't want this step to fail if Amazon is slow.

```yaml
jobs:
  # other jobs...
  build-and-test:
    steps:
      # ... previous steps ...
      - name: Save WarpBuild snapshot
        uses: WarpBuilds/snapshot-save@v1
        if: ${{ steps.find-ddev.outputs.ddev-path != '/usr/bin/ddev' }}
        # Using a matrix build? Avoid thrashing snapshots by only saving from one shard.
        # if: ${{ matrix.shard == 1 && steps.find-ddev.outputs.ddev-path != '/usr/bin/ddev'}}
        with:
          # Must match the snapshot.key in runs-on above

>           alias: "my-project-ddev-1.25.0-v1-${{ needs.determine-snapshot.outputs.snapshot }}"

Reviewer 2: ```suggestion
          alias: "my-project-ddev-1.25.1-v1-${{ needs.determine-snapshot.outputs.snapshot }}"
```

Bump DDEV version
